### PR TITLE
Adding schedule preview page for LineTimetable.

### PR DESCRIPTION
**Pull request by vpassama**
# Schedule preview page

This PR contains the preview page for a LineTimetable schedules data and is related to the issue https://github.com/CanalTP/MttBundle/issues/61.
I'll edit the presentation soon. For the moment, the feature is not finished.

### CODE REVIEW

**Comment by vpassama** [comment]
The branch is updated. So the notes won't be added in this PR.

File: Services/CalendarManager.php [hunk 1]
```diff
@@ -423,8 +423,164 @@ public function isIncluded($calendarId, Season $season)
...
         return false;
     }
 
+    /**
+     * Returning calendars for a line.
+     *
+     * @param string $externalCoverageId
+     * @param string $externalNetworkId
+     * @param string $externalLineId,
+     * @param array  $parameters
```
Comment (datanel) [inline Services/CalendarManager.php]: this param does not exist in the method

Comment (vpassama) [inline Services/CalendarManager.php]: right, it will soon but for now I have to delete this comment.

Comment (datanel) [inline Services/CalendarManager.php]: I think getCalendarsForLine and buildFullCalendar could be unit tested

File: Resources/views/LineTimetable/schedule.html.twig [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+                    <p class="text-center">{{ 'calendar.empty'|trans({}, 'messages') }}</p>
+                </div>
+            {% endfor %}
+        </div>
+    {% endfor %}
+{% endblock %}
+
+{% block javascripts %}
+    <script>
+        $(document).on('change', '#select-route', function() {
```
Comment (datanel) [inline Resources/views/LineTimetable/schedule.html.twig]: I have the followinf error here:
"Uncaught ReferenceError: $ is not defined"

as you don't load jquery with require

```javascript
require(['jquery'], function ($) {
        $(document).on('change', '#select-route', function() {
            $('.calendars-tab').hide();
            $('#'+$(this).val()+'.calendars-tab').show();
        });
        $('#select-route').trigger('change');
});
```

File: Resources/views/LineTimetable/schedule.html.twig [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    {% for routeId, route in schedule %}
+        <ul class="nav nav-tabs calendars-tab" id="{{ loop.index }}">
+            {% for calendar in route.calendars %}
+                <li{% if loop.first %} class="active"{% endif %}>
+                    <a href="#{{ loop.parent.loop.index }}-calendar-{{ calendar.id }}" data-toggle="tab">{{ calendar.name }}</a>
+                </li>
+            {% endfor %}
+        </ul>
+
+        <div class="tab-content calendars-tab" id="{{ loop.index }}">
```
Comment (datanel) [inline Resources/views/LineTimetable/schedule.html.twig]: you have a duplicated id here see the ul element above

Comment (vpassama) [inline Resources/views/LineTimetable/schedule.html.twig]: yes it's used to display the correct route nav tab and associated content.

```
$('#'+$(this).val()+'.calendars-tab').show();
```

Comment (datanel) [inline Resources/views/LineTimetable/schedule.html.twig]: I mean you have the same id here https://github.com/CanalTP/MttBundle/pull/85#discussion-diff-45750488R32

an id is unique

**Comment by datanel** [comment]
tests failed

File: Services/CalendarManager.php [hunk 4]
```diff
@@ -423,8 +423,165 @@ public function isIncluded($calendarId, Season $season)
...
+                    $externalCoverageId,
+                    $externalRouteId,
+                    $calendar->id,
+                    \DateTime::createFromFormat('Ymd', $calendar->validity_pattern->beginning_date)
+                );
+
+                $this->prepareRouteSchedules($routeSchedules, $calendar);
+
+                if (count($routeSchedules->route_schedules) == 0) {
+                    throw new \Exception('No stop points found for the route : '.$externalRouteId);
```
Comment (datanel) [inline Services/CalendarManager.php]: can you add a test for this exception please. This is not covered by your tests

File: Services/CalendarManager.php [hunk 5]
```diff
@@ -423,8 +423,165 @@ public function isIncluded($calendarId, Season $season)
...
+            'metadata' => array()
+        );
 
+        foreach ($schedule->route_schedules as $lineNumber => $stop) {
+            $stopTimes = array();
+            $previousDatetime = null;
+            $dayOffset = false;
+            foreach ($stop->date_times as $columnNumber => $detail) {
+                if (empty($detail->date_time)) {
+                    if (!isset($calendar['metadata'][$columnNumber])) {
```
Comment (datanel) [inline Services/CalendarManager.php]: this condition is not covered by your tests

**Comment by datanel** [comment]
When the attribute column in a calendar is 0 I have the following result (with debug)

line_schedule


you can the table twice. Is is expected?

**Comment by datanel** [comment]
phpunit --coverage-html='path/to/coverage/'

**Comment by vpassama** [comment]
@datanel , the last commit should cover the columns = 0 problem.

File: Resources/views/Layouts/macros.html.twig [hunk 6]
```diff
@@ -206,3 +206,71 @@
...
+    {% if tableNumbers > 0 %}
+        {# this part force the table width to fit with the maximum number of columns #}
+        <tr class="pattern">
+            <th></th>
+            {%- for i in 1..columnsLimit -%}
+                <td>00:00</td>
+            {%- endfor -%}
+        </tr>
+    {% else %}
+        <tr><td><span class="empty-calendar">{{ 'line_timetable.message.empty_calendar'|trans }}</span></td></tr>
```
Comment (datanel) [inline Resources/views/Layouts/macros.html.twig]: this message seems always displayed, even if there is a calendar

In my test calendar.columns = 0 or 1

**Comment by vpassama** [comment]
Is it ok now @datanel ? I'm ready to create next PR and I need this one to be merged. The columns problem is fixed.

File: Services/CalendarManager.php [hunk 7]
```diff
@@ -423,8 +422,181 @@ public function isIncluded($calendarId, Season $season)
...
+            foreach ($calendars as $calendar) {
+                try {
+                    $routeSchedules = $this->navitia->getRouteSchedulesByRouteAndCalendar(
+                        $externalCoverageId,
+                        $externalRouteId,
+                        $calendar->id,
+                        \DateTime::createFromFormat('Ymd', $calendar->validity_pattern->beginning_date)
+                    );
+                } catch(\Exception $e) {
+                    $schedule[$externalRouteId]['calendars'][$calendar->id] = $this->createEmptyLineCalendar($calendar);
```
Comment (datanel) [inline Services/CalendarManager.php]: If you can add a test for this line it would be great and after I can merge ;)

**Comment by vpassama** [comment]
The last test is up.

**Comment by datanel** [comment]
:+1: good job as always

**Pull request merged**